retrieval/discovery/serverset.go

```diff
@@ -327,12 +329,13 @@ func (tc *zookeeperTreeCache) recursiveNodeUpdate(path string, node *zookeeperTr
 				children: map[string]*zookeeperTreeCacheNode{},
 				done:     make(chan struct{}, 1),
 			}
-		}
-		err = tc.recursiveNodeUpdate(path+"/"+child, node.children[child])
-		if err != nil {
-			return err
+			err = tc.recursiveNodeUpdate(path+"/"+child, node.children[child])
```

Comment:
Not sure I follow. Not a Zookeeper expert by any means but doesn't it guarantees that if we delete the whole tree we will see events for all children before the parent event is triggers?